### ISSUE TITLE
System.InvalidCastException in Push when CustomData is empty

### ISSUE THREAD

**Issue by nbsoftware** [Microsoft/AppCenter-SDK-Apple#924]
Hello,

With V1.2.0 (and at least on iOS using Xamarin) I get this exception when CustomData is empty.
Does work fine if CustomData is at least set with one key value pair.
From what I can recall it used to work fine prior to V1.2.0.

```
{System.InvalidCastException: Unable to cast object of type 'Foundation.NSDictionary' to type 'Foundation.NSDictionary`2[[Foundation.NSString, Xamarin.iOS, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065],[Foundation.NSString, Xamarin.iOS, Ver…}
```
Stack trace is : 

```
ObjCRuntime.Runtime.GetNSObject<Foundation.NSDictionary<Foundation.NSString,Foundation.NSString>>( ptr) in /Users/builder/data/lanes/5665/db807ec9/source/xamarin-macios/src/ObjCRuntime/Runtime.cs:1136
Microsoft.AppCenter.Push.iOS.Bindings.MSPushNotification.get_CustomData() in 
Microsoft.AppCenter.Push.Push.<>c.<.cctor>b__2_0(Microsoft.AppCenter.Push.iOS.Bindings.MSPushNotification notification) in 
Microsoft.AppCenter.Push.iOS.PushDelegate.ReceivedPushNotification(Microsoft.AppCenter.Push.iOS.Bindings.MSPush push, Microsoft.AppCenter.Push.iOS.Bindings.MSPushNotification pushNotification) in 
UIKit.UIApplication.UIApplicationMain() in 
UIKit.UIApplication.Main( args, principal, delegate) in /Users/builder/data/lanes/5665/db807ec9/source/xamarin-macios/src/UIKit/UIApplication.cs:79
UIKit.UIApplication.Main( args, principalClassName, delegateClassName) in /Users/builder/data/lanes/5665/db807ec9/source/xamarin-macios/src/UIKit/UIApplication.cs:63
fipacapp.iOS.Application.Main(string[] args) in /work/repositories/customers/fipac/fipac-mobile/fipacapp/fipacapp.iOS/Main.cs:17
```

**Comment by TroubleMakerBen** [comment]
Hey @nbsoftware,

thx for letting us know about this. We're looking into it.
Best,
Benjamin

**Comment by nbsoftware** [comment]
Hi Ivan,

Xamarin iOS is:
Version: 11.6.1.4 (Visual Studio Community)
Hash: db807ec9
Branch: xcode9.2
Build date: 2018-01-10 16:45:48-0500

Hard for me to extract a sample project.
In 1.2.0 it was definitely happening all the time, using the AppCenter UI to push a message.
I will retry this with the new 1.3.0 AppCenter packages and let you know the outcome.

**Comment by MatkovIvan** [comment]
In addition, I'm wondering if you tried just clean build (remove `obj` and `bin` folders)? Sometimes this helps with some strange problems.

**Comment by TroubleMakerBen** [comment]
Hey @nbsoftware,
Any chance that you were able to try the new 1.3.0 packages? 

Bestm
Benjamin

**Comment by nbsoftware** [comment]
Yep.
Still crashes even with 1.3 and full rebuild.

Crash catched by appcenter itself:

```Application.Main (System.String[] args)
System.InvalidCastException: Unable to cast object of type 'Foundation.NSDictionary' to type 'Foundation.NSDictionary`2[[Foundation.NSString, Xamarin.iOS, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065],[Foundation.NSString, Xamarin.iOS, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065]]'
Unable to cast object of type 'Foundation.NSDictionary' to type 'Foundation.NSDictionary`2[[Foundation.NSString, Xamarin.iOS, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065],[Foundation.NSString, Xamarin.iOS, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065]]'
Runtime.GetNSObject[T] (System.IntPtr ptr)
MSPushNotification.get_CustomData ()
<.cctor>b__2_0 (Microsoft.AppCenter.Push.iOS.Bindings.MSPushNotification notification)
PushDelegate.ReceivedPushNotification (Microsoft.AppCenter.Push.iOS.Bindings.MSPush push, Microsoft.AppCenter.Push.iOS.Bindings.MSPushNotification pushNotification)
(wrapper managed-to-native) UIKit.UIApplication:UIApplicationMain (int,string[],intptr,intptr)
UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate)
UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName)
Application.Main (System.String[] args)
```

**Comment by TroubleMakerBen** [comment]
@nbsoftware we're looking into this. Could we sent you a pre-release binary to test?

**Comment by nbsoftware** [comment]
Of course - provided that I know how to integrate them if I cannot retrieve those from nuget.

Also, I realize that this issue should probably have been posted in https://github.com/Microsoft/AppCenter-SDK-DotNet rather than here...

**Comment by TroubleMakerBen** [comment]
Hey Frederic,

We've looked at this issue in detail and we're checking for nil/null and types in all appropriate places and as the issue happens in generated code, there isn't much we can do, especially as we cannot repro this issue.
As you mentioned that it's hard to extract a sample, could you use our sample project (which is OSS) and repro it on your machine using that project? That way, we could try to help you figure out what's going on?
1. Clone [our repo](https://github.com/Microsoft/AppCenter-SDK-Dotnet) and checkout the master branch
2. Open `AppCenter-Mac.sln`
3. Use Contoso.Forms.Puppet.iOS. You'd need to use your app secret to setup your SDK, change to your bundle identifier and use your certs/provisioning profile for signing.

Best,
Benjamin

PS: in case you check out our sample project, maybe you can compare with your project setup and see if there's a difference in config that explains the issue you are seeing?

**Comment by TroubleMakerBen** [comment]
Hey! 

We haven't heard from you in a while, so I am going to go ahead and close this issue. Should you still have any questions, please don't hesitate to contact us again. 

Thanks for using App Center!

**Comment by Swiftless** [comment]
I can confirm I am having the exact same issue with Microsoft.AppCenter.Push v1.5.0

Upon receiving push notifications while having my iOS App open, it crashed after receiving every notification. My device receives the notifications fine when the app is closed and I can open my app from the notifications without any issue.

The logs in AppCenter state:

> System.InvalidCastException: Unable to cast object of type 'Foundation.NSDictionary' to type 'Foundation.NSDictionary`2[[Foundation.NSString, Xamarin.iOS, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065],[Foundation.NSString, Xamarin.iOS, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065]]'
> Runtime.GetNSObject[T] (System.IntPtr ptr)
> MSPushNotification.get_CustomData ()
> <.cctor>b__2_0 (Microsoft.AppCenter.Push.iOS.Bindings.MSPushNotification notification)
> PushDelegate.ReceivedPushNotification (Microsoft.AppCenter.Push.iOS.Bindings.MSPush push, Microsoft.AppCenter.Push.iOS.Bindings.MSPushNotification pushNotification)
> (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
> UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) /Library/Frameworks/Xamarin.iOS.framework/Versions/11.9.1.24/src/Xamarin.iOS/UIKit/UIApplication.cs:79
> UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) /Library/Frameworks/Xamarin.iOS.framework/Versions/11.9.1.24/src/Xamarin.iOS/UIKit/UIApplication.cs:63

**Comment by TroubleMakerBen** [comment]
Hi @Swiftless,

can you get in touch using Intercom (blue button on app center portal) so we can discuss this in more detail? I'd love to see the full symbolicated crash log and verbose loggin enabled.

**Comment by Swiftless** [comment]
Yes of course. I'll do that now. I will be finishing up for the day within the next 30 minutes or so.

**Comment by guperrot** [comment]
Hey! 

We haven’t heard from you in a while, so I am going to go ahead and close this ticket. Should you still have any questions, you can write a message here & that will reopen the ticket; or if you prefer you can create a new ticket. Please don’t hesitate to contact us again.

Thanks for using App Center!

**Comment by Swiftless** [comment]
Hi @guperrot,

Through conversation and testing with @TroubleMakerBen this has been fixed in Version 1.6.0 on April 20th with the release note of:

> [Fix] Fix parsing custom data for push to avoid crash in the application.

Thanks!

**Comment by guperrot** [comment]
Hi, thanks for pointing that out, you are right about the release notes.
Thanks for using App Center.
